File: backend/DataAccess/Common/Models/DaSelectorModel.cs
// *******************************************************************
// 業務名称　: 健康管理システム
// 機能概要　: ドロップダウンリスト
//
// 作成日　　: 2023.03.14
// 作成者　　: 蔣
// 変更履歴　:
// *******************************************************************
using static BCC.Affect.DataAccess.DaConvertUtil;

namespace BCC.Affect.DataAccess
{
    public class DaSelectorModel
    {
        public string value { get; set; }   //コード
        public string label { get; set; }   //名称

        /// <summary>
        /// Cacheに使用
        /// </summary>
        public DaSelectorModel() { }

        public DaSelectorModel(string? value, string? label)
        {
            this.value = value ?? string.Empty;
            this.label = label ?? string.Empty;
        }

        public override string ToString()
        {
            return $"{value}{DaConst.SELECTOR_DELIMITER}{label}";
        }
    }
    public class DaSelectorKeyModel : DaSelectorModel
    {
        public string key { get; set; }   //キー項目(連動フィルター用)
        public DaSelectorKeyModel(string? value, string? label, string? key) : base(value, label)
        {
            this.key = key ?? string.Empty;
        }
    }

    public class DaSelectorDisabledModel : DaSelectorModel
    {
        public bool disabled { get; set; }   //無効属性
        public DaSelectorDisabledModel(string? value, string? label, bool? disabled = false) : base(value, label)
        {
            this.disabled = CBool(disabled);
        }
    }

    public class DaSelectorTreeModel<T> : DaSelectorModel where T : DaSelectorModel
    {
        public List<T> children { get; set; } //サブレベルオプション

        public DaSelectorTreeModel(string? value, string? label, IEnumerable<T>? children = null) : base(value, label)
        {
            this.children = children?.ToList() ?? new List<T>();
        }
    }
}

File: backend/JbdGjsService/DataAccess/Common/Models/DaSelectorModel.vb
' *******************************************************************
' 業務名称　: 互助事業システム
' 機能概要　: ドロップダウンリスト
'
' 作成日　　: 2024.07.14
' 作成者　　: 
' 変更履歴　:
' *******************************************************************

Namespace JBD.GJS.Service
    Public Class DaSelectorModel
        Public Property value As String   'コード
        Public Property label As String   '名称

        ''' <summary>
        ''' Cacheに使用
        ''' </summary>
        Public Sub New()
        End Sub

        Public Sub New(value As String, label As String)
            Me.value = If(value, String.Empty)
            Me.label = If(label, String.Empty)
        End Sub

        Public Overrides Function ToString() As String
            Return $"{value}{DaConst.SELECTOR_DELIMITER}{label}"
        End Function
    End Class
    Public Class DaSelectorKeyModel
        Inherits DaSelectorModel
        Public Property key As String   'キー項目(連動フィルター用)
        Public Sub New(value As String, label As String, key As String)
            MyBase.New(value, label)
            Me.key = If(key, String.Empty)
        End Sub
    End Class

    Public Class DaSelectorDisabledModel
        Inherits DaSelectorModel
        Public Property disabled As Boolean   '無効属性
        Public Sub New(value As String, label As String, Optional disabled As Boolean = False)
            MyBase.New(value, label)
            Me.disabled = DaConvertUtil.CBool(disabled)
        End Sub
    End Class

    Public Class DaSelectorTreeModel(Of T As DaSelectorModel)
        Inherits DaSelectorModel
        Public Property children As List(Of T) 'サブレベルオプション

        Public Sub New(value As String, label As String, Optional children As IEnumerable(Of T) = Nothing)
            MyBase.New(value, label)
            Me.children = If(children.ToList(), New List(Of T)())
        End Sub
    End Class
End Namespace
